server/src/main/java/com/ncc/neon/services/ClusterService.java

```diff
@@ -23,16 +26,190 @@
+                LinkedHashMap map = new LinkedHashMap<>();
+                ArrayList range = new ArrayList<>();
+                if (i != 0) {
+                    range.add(currentBin.add(step));
+                } else {
+                    range.add(currentBin);
+                }
+                range.add(currentBin = currentBin.add(gap));
+                map.put(fieldNameKey, range);
+                newData.add(map);
+            }
+        } else {
+            for (int i = 0; i < clusters.size(); i++) {
+                LinkedHashMap map = new LinkedHashMap<>();
+                ArrayList range = new ArrayList();
+                range.add(new BigDecimal(clusters.get(i).get(0).toString()));
+                range.add(new BigDecimal(clusters.get(i).get(1).toString()));
+                map.put(fieldNameKey, range);
+                newData.add(map);
+            }
+        }
+        return newData;
+    }
+
+    /**
+     * Setup any extra keys
+     *
+     * @param fieldNameKey fieldname key given by the cluster clause
+     * @param aggregationNameKey aggregation name key given by the cluster clause
+     * @param data original result data
+     * @return the extra key sets map for storing additional keys
+     */
+    private Map<String, Object> getExtraKeySetsMap(String fieldNameKey, String aggregationNameKey,
+                                                   List<Map<String, Object>> data) {
+        Map<String, Object> datum = data.get(0);
+        Map<String, Object> extraKeySets = new HashMap<>();
+        for (String key : datum.keySet()) {
+            if (!key.equals(aggregationNameKey) && !key.equals(fieldNameKey)) {
+                extraKeySets.put(key, new HashSet<>());
+            }
+        }
+        return extraKeySets;
+    }
+
+    /**
+     * Traverses the original data and puts it in the correct bins in the new data.
+     *
+     * @param fieldNameKey the key that corresponds with the field name
+     * @param aggregationNameKey the key that corresponds with the overall counts of each bin
+     * @param data the original result data
+     * @param newData the newly clustered data
+     * @param extraKeySets the
+     */
+    private void aggregateNumbersInNewData(String fieldNameKey, String aggregationNameKey, List<Map<String,
+            Object>> data, List<Map<String, Object>> newData, Map<String, Object> extraKeySets) {
+        Iterator<Map<String, Object>> newDataIter = newData.iterator();
+        int oldDataIndex = 0;
+        while (newDataIter.hasNext()) {
+            // reset the extra keys for each new bin
+            for (String key : extraKeySets.keySet()) {
+                extraKeySets.put(key, new HashSet<>());
+            }
+
+            // determine the boundaries and aggregated count for this bin
+            BigDecimal currAgg = new BigDecimal("0");
+            Map currNewBin = newDataIter.next();
+            ArrayList alo = (ArrayList) currNewBin.get(fieldNameKey);
```

Comment:
What does `alo` mean?

Reply:
ArrayListObject, a genuinely terrible variable name...

Reply:
I renamed it 😆